File: python/tests/test_error_conditions.py
"""
Error Condition Tests (ERR family)

Gould (2001): "Test that your code does not work in circumstances where it should not."
These tests verify that invalid inputs are rejected with clear error messages.

No network access required.
"""

import pytest
import pandas as pd
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent.parent
FIXTURES = REPO_ROOT / "tests" / "fixtures" / "deterministic"


# ===========================================================================
# ERR-04: Invalid country code
# ===========================================================================

class TestInvalidCountryCode:
    """Invalid ISO3 codes should be rejected or warned."""

    def test_too_short_code_rejected(self):
        from unicefdata.utils import validate_country_codes
        with pytest.raises(ValueError):
            validate_country_codes(["US"])

    def test_lowercase_code_rejected(self):
        from unicefdata.utils import validate_country_codes
        with pytest.raises(ValueError):
            validate_country_codes(["usa"])

    def test_non_string_rejected(self):
        from unicefdata.utils import validate_country_codes
        with pytest.raises(ValueError):
            validate_country_codes([123])

    def test_valid_codes_pass(self):
        from unicefdata.utils import validate_country_codes
        result = validate_country_codes(["USA", "BRA"])
        assert "USA" in result
        assert "BRA" in result

    def test_format_valid_unknown_passes(self):
        """ZZZ is format-valid (3 uppercase) — accepted without valid_codes set."""
        from unicefdata.utils import validate_country_codes
        result = validate_country_codes(["ZZZ"])
        assert "ZZZ" in result

    def test_unknown_rejected_with_valid_set(self):
        """ZZZ rejected when valid_codes set is provided."""
        from unicefdata.utils import validate_country_codes
        with pytest.raises(ValueError):
            validate_country_codes(["ZZZ"], valid_codes={"USA", "BRA"})


# ===========================================================================
# ERR-05: Inverted year range
# ===========================================================================

class TestInvertedYearRange:
    """Year range 2025:2020 should be rejected."""

    def test_inverted_range_raises(self):
        from unicefdata.utils import validate_year_range
        with pytest.raises(ValueError):
            validate_year_range(2025, 2020)

    def test_valid_range_passes(self):
        from unicefdata.utils import validate_year_range
        result = validate_year_range(2015, 2023)
        assert result is not None


# ===========================================================================
# ERR-06: No indicator provided
# ===========================================================================

class TestNoIndicator:
    """Calling unicefData without an indicator should raise."""

    def test_none_indicator(self):
        from unicefdata import unicefData
        with pytest.raises((TypeError, ValueError)):
            unicefData(indicator=None)

    def test_empty_string_indicator(self):
        from unicefdata import unicefData
        with pytest.raises((ValueError, Exception)):
            unicefData(indicator="")


# ===========================================================================
# ERR-08: Invalid indicator code
# ===========================================================================

class TestInvalidIndicator:
    """Invalid indicator should raise SDMXNotFoundError (requires some metadata)."""

    def test_invalid_indicator_metadata_lookup(self):
        """get_dataflow_for_indicator returns None for unknown indicators."""
        from unicefdata.indicator_registry import get_dataflow_for_indicator
        result = get_dataflow_for_indicator("XXXXX_INVALID_99999")
        # Should return None or raise — both are acceptable
        assert result is None or isinstance(result, str)


# ===========================================================================
# Year parsing validation
# ===========================================================================

class TestYearParsing:
    """parse_year() should handle various input formats correctly."""

    def test_single_int(self):
        from unicefdata.unicefdata import parse_year
        result = parse_year(2020)
        assert result["start_year"] == 2020
        assert result["end_year"] == 2020

    def test_range_string(self):
        from unicefdata.unicefdata import parse_year
        result = parse_year("2015:2023")
        assert result["start_year"] == 2015
        assert result["end_year"] == 2023

    def test_comma_string(self):
        from unicefdata.unicefdata import parse_year
        result = parse_year("2015,2018,2020")
        assert result["year_list"] == [2015, 2018, 2020]

    def test_none_returns_none(self):
        from unicefdata.unicefdata import parse_year
        result = parse_year(None)
        assert result["start_year"] is None
        assert result["end_year"] is None

    def test_tuple_range(self):
        from unicefdata.unicefdata import parse_year
        result = parse_year((2015, 2023))
        assert result["start_year"] == 2015
        assert result["end_year"] == 2023


# ===========================================================================
# Duplicate detection (DL-06)
# ===========================================================================

class TestDuplicateDetection:
    """Fixtures should not have duplicate rows on key dimensions."""

    @pytest.mark.parametrize("fixture", [
        "CME_MRY0T4_USA_2020_pinning.csv",
        "CME_MRY0T4_USA_BRA_2020.csv",
        "CME_MRY0T4_USA_2015_2023.csv",
    ])
    def test_no_duplicates_on_key_dims(self, fixture):
        df = pd.read_csv(FIXTURES / fixture)
        key_cols = [c for c in ["REF_AREA", "INDICATOR", "SEX", "WEALTH_QUINTILE", "TIME_PERIOD"]
                    if c in df.columns]
        duplicates = df.duplicated(subset=key_cols, keep=False)
        assert not duplicates.any(), f"Duplicate rows found in {fixture}"


# ===========================================================================
# Data type validation (DATA-01)
# ===========================================================================

class TestDataTypes:
    """Numeric columns must be parseable as numbers."""

    @pytest.mark.parametrize("fixture", [
        "CME_MRY0T4_USA_2020_pinning.csv",
        "CME_MRY0T4_BRA_sex_2020.csv",
        "CME_multi_USA_2020.csv",
        "IM_MCV1_USA_BRA_2015_2023.csv",
    ])
    def test_obs_value_numeric(self, fixture):
        df = pd.read_csv(FIXTURES / fixture)
        assert pd.to_numeric(df["OBS_VALUE"], errors="coerce").notna().all(), \
            f"OBS_VALUE has non-numeric values in {fixture}"

    @pytest.mark.parametrize("fixture", [
        "CME_MRY0T4_USA_2020_pinning.csv",
        "CME_MRY0T4_USA_2015_2023.csv",
        "CME_MRY0T4_BRA_1990_2023.csv",
    ])
    def test_time_period_integer(self, fixture):
        df = pd.read_csv(FIXTURES / fixture)
        periods = pd.to_numeric(df["TIME_PERIOD"], errors="coerce")
        assert periods.notna().all()
        assert (periods == periods.astype(int)).all(), \
            f"TIME_PERIOD should be integer years in {fixture}"

    @pytest.mark.parametrize("fixture", [
        "CME_MRY0T4_USA_2020_pinning.csv",
        "CME_MRY0T4_USA_BRA_2020.csv",
        "CME_MRY0T4_multi_2018_2023.csv",
    ])
    def test_iso3_three_chars(self, fixture):
        df = pd.read_csv(FIXTURES / fixture)
        assert (df["REF_AREA"].str.len() == 3).all(), \
            f"REF_AREA should be 3-char ISO3 codes in {fixture}"

File: tests/testthat/test-error-conditions.R
# =============================================================================
# Error Condition Tests (ERR family)
#
# Gould (2001): "Test that your code does not work in circumstances where it
# should not." Validates parameter rejection and graceful error handling.
#
# No network access required (except ERR-04/08 which test API error paths).
# Uses helper-fixtures.R for path resolution (auto-loaded by testthat)
# =============================================================================

# FIXTURES_DIR provided by helper-fixtures.R
FIXTURES_DIR <- get_fixtures_dir()

# ===========================================================================
# Year parsing validation
# ===========================================================================

test_that("parse_year handles single integer", {
  skip_if_not_installed("unicefData")
  result <- unicefData:::parse_year(2020)
  expect_equal(result$start_year, 2020)
  expect_equal(result$end_year, 2020)
})

test_that("parse_year handles range string", {
  skip_if_not_installed("unicefData")
  result <- unicefData:::parse_year("2015:2023")
  expect_equal(result$start_year, 2015)
  expect_equal(result$end_year, 2023)
})

test_that("parse_year handles comma-separated string", {
  skip_if_not_installed("unicefData")
  result <- unicefData:::parse_year("2015,2018,2020")
  expect_equal(result$year_list, c(2015, 2018, 2020))
})

test_that("parse_year handles NULL", {
  skip_if_not_installed("unicefData")
  result <- unicefData:::parse_year(NULL)
  expect_null(result$start_year)
  expect_null(result$end_year)
})

# ===========================================================================
# DL-06: Duplicate detection
# ===========================================================================

test_that("DL-06: pinning fixture has no duplicates on key dims", {
  df <- read.csv(file.path(FIXTURES_DIR, "CME_MRY0T4_USA_2020_pinning.csv"),
                 stringsAsFactors = FALSE)
  key_cols <- intersect(
    c("REF_AREA", "INDICATOR", "SEX", "WEALTH_QUINTILE", "TIME_PERIOD"),
    names(df)
  )
  n_total <- nrow(df)
  n_unique <- nrow(unique(df[key_cols]))
  expect_equal(n_total, n_unique, info = "No duplicate rows on key dimensions")
})

test_that("DL-06: multi-country fixture has no duplicates", {
  df <- read.csv(file.path(FIXTURES_DIR, "CME_MRY0T4_USA_BRA_2020.csv"),
                 stringsAsFactors = FALSE)
  key_cols <- intersect(
    c("REF_AREA", "INDICATOR", "SEX", "WEALTH_QUINTILE", "TIME_PERIOD"),
    names(df)
  )
  expect_equal(nrow(df), nrow(unique(df[key_cols])))
})

test_that("DL-06: time series fixture has no duplicates", {
  df <- read.csv(file.path(FIXTURES_DIR, "CME_MRY0T4_USA_2015_2023.csv"),
                 stringsAsFactors = FALSE)
  key_cols <- intersect(
    c("REF_AREA", "INDICATOR", "SEX", "WEALTH_QUINTILE", "TIME_PERIOD"),
    names(df)
  )
  expect_equal(nrow(df), nrow(unique(df[key_cols])))
})

# ===========================================================================
# DATA-01: Data type validation
# ===========================================================================

test_that("DATA-01: OBS_VALUE is numeric across fixtures", {
  fixtures <- c("CME_MRY0T4_USA_2020_pinning.csv",
                "CME_MRY0T4_BRA_sex_2020.csv",
                "CME_multi_USA_2020.csv",
                "IM_MCV1_USA_BRA_2015_2023.csv")
  for (f in fixtures) {
    df <- read.csv(file.path(FIXTURES_DIR, f), stringsAsFactors = FALSE)
    expect_true(is.numeric(df$OBS_VALUE), info = paste(f, ": OBS_VALUE not numeric"))
  }
})

test_that("DATA-01: TIME_PERIOD is integer years", {
  fixtures <- c("CME_MRY0T4_USA_2020_pinning.csv",
                "CME_MRY0T4_USA_2015_2023.csv",
                "CME_MRY0T4_BRA_1990_2023.csv")
  for (f in fixtures) {
    df <- read.csv(file.path(FIXTURES_DIR, f), stringsAsFactors = FALSE)
    expect_true(is.numeric(df$TIME_PERIOD), info = paste(f, ": TIME_PERIOD not numeric"))
    expect_true(all(df$TIME_PERIOD == as.integer(df$TIME_PERIOD)),
                info = paste(f, ": TIME_PERIOD not integer"))
  }
})

test_that("DATA-01: REF_AREA is 3-character ISO3", {
  fixtures <- c("CME_MRY0T4_USA_2020_pinning.csv",
                "CME_MRY0T4_USA_BRA_2020.csv",
                "CME_MRY0T4_multi_2018_2023.csv")
  for (f in fixtures) {
    df <- read.csv(file.path(FIXTURES_DIR, f), stringsAsFactors = FALSE)
    expect_true(all(nchar(df$REF_AREA) == 3),
                info = paste(f, ": REF_AREA not 3 chars"))
  }
})

# ===========================================================================
# ERR-06: No indicator provided
# ===========================================================================

test_that("ERR-06: unicefData with no indicator errors", {
  skip_if_not_installed("unicefData")
  expect_error(unicefData::unicefData())
})

# ===========================================================================
# ERR: Mutually exclusive format options
# ===========================================================================

test_that("ERR: empty indicator string is handled", {
  skip_if_not_installed("unicefData")
  # Empty string should not succeed silently
  result <- tryCatch(
    unicefData::unicefData(indicator = ""),
    error = function(e) "error"
  )
  expect_true(result == "error" || (is.data.frame(result) && nrow(result) == 0))
})
